### ISSUE TITLE
Support GPU/CPU communication in RPC

### ISSUE THREAD

**Issue by rafi-cohen** [pytorch/pytorch#53696]
## 🚀 Feature
PyTorch 1.8 introduces CUDA-support in RPC. This is done by simply specifying device mappings between workers.
However, device maps can only be defined between CUDA devices:

https://github.com/pytorch/pytorch/blob/215950e2be77398adec7371e3f97e8979d4d791a/torch/distributed/rpc/options.py#L103-L109

I would like to request support for defining device maps between CUDA devices and CPU.

## Motivation

I'm trying to implement a CPU parameter server. The workers compute gradients using GPUs and then send them to the parameter server. The parameter server computes the updated parameters (on CPU) and sends them back to the worker.
Supporting GPU/CPU communication directly could provide speed improvements by saving time on copying.

## Pitch

Defining a device map from a GPU to CPU or vice-versa should be supported. For example (based on the one provided in the documentation):
```python
# both workers
def add(x, y):
    print(x)  # tensor([1., 1.])  (device='cpu')
    return x + y, (x + y).to(2)

# on worker 0
options = TensorPipeRpcBackendOptions(
    num_worker_threads=8,
    device_maps={"worker1": {0: 'cpu'}}
    # maps worker0's cuda:0 to worker1's cpu
)
options.set_device_map("worker1", {'cpu': 2})
# maps worker0's cpu to worker1's cuda:2

rpc.init_rpc(
    "worker0",
    rank=0,
    world_size=2,
    backend=rpc.BackendType.TENSORPIPE,
    rpc_backend_options=options
)

x = torch.ones(2)
rets = rpc.rpc_sync("worker1", add, args=(x.to(0), 1))
# The first argument will be moved to cpu on worker1. When
# sending the return value back, it will follow the invert of
# the device map, and hence will be moved back to cuda:0 and
# cpu on worker0
print(rets[0])  # tensor([2., 2.], device='cuda:0')
print(rets[1])  # tensor([2., 2.])  (device='cpu')
```
## Alternatives

Currently, the only way I'm aware of to send from a GPU to a CPU using RPC is by first copying the GPU tensors to the CPU and then send them over, e.g.:
```python
rets = rpc.rpc_sync("worker_name", some_func, args=(gpu_tensor.to('cpu'), ...))
```
Of course, copying can be performed beforehand, preferably in a non-blocking way.

## Additional context

Follow up to #53501

**Comment by mrshenli** [comment]
cc @beauby

**Comment by beauby** [comment]
@rafi-cohen Thanks for your interest, we are currently working on supporting this in TensorPipe.

**Comment by weberxie** [comment]
@beauby Could you please update the progress and the available time on supporting this feature? Thanks!
